File: src/calibration.py
"""Auditoría de calibración para StrikeoutLab.

Este módulo es la razón de ser del proyecto: mide si las confianzas
asignadas a los picks (calculadas o de juicio) se sostienen contra los
resultados reales. Si después de suficientes picks la calibración muestra
que no se sostienen, ese es un resultado válido del sistema, no un fallo.
"""
from __future__ import annotations

import pandas as pd

BANDAS = [
    (0.70, 0.75, "70-74%"),
    (0.75, 0.80, "75-79%"),
    (0.80, 0.85, "80-84%"),
    (0.85, 0.90, "85-89%"),
    (0.90, 0.95, "90-94%"),
    (0.95, 1.01, "95-99%"),  # límite superior inclusivo de 1.0 (100%)
]

MUESTRA_MINIMA = 20

RESULTADOS_RESUELTOS = ("GANO", "PERDIO", "EMPATE")


def _banda_de(confianza: float) -> str | None:
    for lo, hi, etiqueta in BANDAS:
        if lo <= confianza < hi:
            return etiqueta
    return None


def _resumir_grupo(grupo: pd.DataFrame) -> dict:
    ganadas = int((grupo["resultado"] == "GANO").sum())
    perdidas = int((grupo["resultado"] == "PERDIO").sum())
    empates = int((grupo["resultado"] == "EMPATE").sum())
    decididas = ganadas + perdidas
    tasa_real = (ganadas / decididas) if decididas else None
    confianza_promedio = float(grupo["confianza"].mean())
    diferencia = (confianza_promedio - tasa_real) if tasa_real is not None else None

    return {
        "cantidad": len(grupo),
        "ganadas": ganadas,
        "perdidas": perdidas,
        "empates": empates,
        "confianza_promedio": confianza_promedio,
        "tasa_real": tasa_real,
        "diferencia": diferencia,
        "muestra_insuficiente": len(grupo) < MUESTRA_MINIMA,
    }


def reporte_calibracion(picks: pd.DataFrame) -> pd.DataFrame:
    """Agrupa los picks resueltos en bandas de confianza y compara la
    confianza promedio declarada contra la tasa real de acierto.

    Interpretación esperada: si una banda muestra tasa_real muy por debajo
    de confianza_promedio (diferencia positiva grande), el sistema está
    sobreconfiado en esa banda y las confianzas deben ajustarse a la baja.

    Los empates cuentan en 'cantidad' pero se excluyen del denominador de
    tasa_real (ganadas / (ganadas + perdidas)): un push no confirma ni
    refuta la confianza asignada.

    Cada banda se desglosa también por fuente_confianza (fila 'TODAS',
    'CALCULADA' y 'JUICIO'), porque mezclar ambas sin distinguirlas fue el
    error original que este proyecto corrige.

    Una banda con menos de 20 picks queda marcada
    muestra_insuficiente=True; no debe usarse para concluir que esa banda
    está sobre o subconfiada.
    """
    requeridas = {"resultado", "confianza", "fuente_confianza"}
    faltantes = requeridas - set(picks.columns)
    if faltantes:
        raise ValueError(f"picks no tiene las columnas requeridas: {faltantes}")

    resueltos = picks[picks["resultado"].isin(RESULTADOS_RESUELTOS)].copy()
    resueltos["banda"] = resueltos["confianza"].apply(_banda_de)
    resueltos = resueltos.dropna(subset=["banda"])

    filas = []
    for _, _, etiqueta in BANDAS:
        banda_df = resueltos[resueltos["banda"] == etiqueta]
        if banda_df.empty:
            continue

        filas.append({"banda": etiqueta, "fuente_confianza": "TODAS", **_resumir_grupo(banda_df)})

        for fuente in ("CALCULADA", "JUICIO"):
            sub_df = banda_df[banda_df["fuente_confianza"] == fuente]
            if sub_df.empty:
                continue
            filas.append({"banda": etiqueta, "fuente_confianza": fuente, **_resumir_grupo(sub_df)})

    columnas = [
        "banda", "fuente_confianza", "cantidad", "ganadas", "perdidas",
        "empates", "confianza_promedio", "tasa_real", "diferencia",
        "muestra_insuficiente",
    ]
    return pd.DataFrame(filas, columns=columnas)


def resumen_economico(picks: pd.DataFrame) -> dict:
    """Resultado económico real de los picks resueltos, sin adornos.

    El desglose 'por_nivel' reporta cantidad y resultados (GANO/PERDIO/
    EMPATE) por nivel de pureza (Diamante/Oro/etc.), no montos: una misma
    boleta física puede combinar patas de distinto nivel, así que repartir
    el stake/payout de esa boleta entre niveles sería arbitrario.

    Los montos totales (stake/payout) solo se calculan si picks.csv incluye
    esas columnas opcionales. Cuando varias patas comparten 'ticket_id' se
    asume que el stake/payout de la boleta física está registrado de forma
    idéntica en cada una de sus filas, así que se deduplica por ticket_id
    antes de sumar para no contar el monto de una misma boleta más de una
    vez. Filas sin ticket_id se tratan como apuestas sueltas.
    """
    resueltos = picks[picks["resultado"].isin(RESULTADOS_RESUELTOS)]

    resumen: dict = {"total_picks_resueltos": len(resueltos), "por_nivel": {}}

    for nivel, grupo in resueltos.groupby("nivel"):
        resumen["por_nivel"][nivel] = {
            "cantidad": len(grupo),
            "ganadas": int((grupo["resultado"] == "GANO").sum()),
            "perdidas": int((grupo["resultado"] == "PERDIO").sum()),
            "empates": int((grupo["resultado"] == "EMPATE").sum()),
        }

    tiene_columnas_dinero = "stake" in picks.columns and "payout" in picks.columns
    dinero = pd.DataFrame()
    if tiene_columnas_dinero:
        con_ticket = resueltos[resueltos["ticket_id"].notna()].drop_duplicates(
            subset=["ticket_id"]
        )
        sin_ticket = resueltos[resueltos["ticket_id"].isna()]
        dinero = pd.concat([con_ticket, sin_ticket])

    # Un hueco explícito es preferible a un número inventado: si no hay ni
    # una fila con stake/payout realmente registrado, reportar None en vez
    # de sumar puros NaN y mostrar un engañoso "0.00" que parecería un
    # resultado económico real.
    tiene_datos_reales = tiene_columnas_dinero and (
        dinero["stake"].notna().any() or dinero["payout"].notna().any()
    )

    if tiene_datos_reales:
        total_apostado = float(dinero["stake"].sum())
        total_cobrado = float(dinero["payout"].sum())

        resumen["total_apostado"] = total_apostado
        resumen["total_cobrado"] = total_cobrado
        resumen["neto"] = total_cobrado - total_apostado
    else:
        resumen["total_apostado"] = None
        resumen["total_cobrado"] = None
        resumen["neto"] = None
        if tiene_columnas_dinero:
            resumen["advertencia"] = (
                "Ningún pick resuelto tiene 'stake'/'payout' registrado; no "
                "se puede calcular el resultado económico en pesos."
            )
        else:
            resumen["advertencia"] = (
                "picks.csv no tiene columnas 'stake'/'payout'; no se puede "
                "calcular el resultado económico en pesos."
            )

    return resumen

File: packages/core/src/calibration.ts
/**
 * Auditoría de calibración de StrikeoutLab.
 *
 * Este módulo es la razón de ser del proyecto: mide si las confianzas
 * asignadas a los picks (CALCULADA o JUICIO) se sostienen contra los
 * resultados reales. También es la fuente del "ciclo de retroalimentación"
 * que las Edge Functions le pasan como contexto a la IA — no es que el
 * modelo se reentrena, es que ve su propio historial real antes de opinar.
 */

import type { Resultado } from "./calculations.js";

export type FuenteConfianza = "CALCULADA" | "JUICIO";
export type Nivel = "DIAMANTE" | "ORO_ALTO" | "ORO" | "IMPUREZA";

export interface PickCalibracion {
  confianza: number;
  resultado: Resultado | null;
  fuenteConfianza: FuenteConfianza;
}

export interface PickEconomico {
  resultado: Resultado | null;
  nivel: Nivel;
  ticketId?: string | null;
  stake?: number | null;
  payout?: number | null;
}

export interface BandaCalibracion {
  banda: string;
  fuenteConfianza: "TODAS" | FuenteConfianza;
  cantidad: number;
  ganadas: number;
  perdidas: number;
  empates: number;
  confianzaPromedio: number;
  tasaReal: number | null;
  diferencia: number | null;
  muestraInsuficiente: boolean;
}

export interface ResumenEconomico {
  totalPicksResueltos: number;
  porNivel: Record<string, { cantidad: number; ganadas: number; perdidas: number; empates: number }>;
  totalApostado: number | null;
  totalCobrado: number | null;
  neto: number | null;
  advertencia?: string;
}

const BANDAS: Array<{ lo: number; hi: number; etiqueta: string }> = [
  { lo: 0.7, hi: 0.75, etiqueta: "70-74%" },
  { lo: 0.75, hi: 0.8, etiqueta: "75-79%" },
  { lo: 0.8, hi: 0.85, etiqueta: "80-84%" },
  { lo: 0.85, hi: 0.9, etiqueta: "85-89%" },
  { lo: 0.9, hi: 0.95, etiqueta: "90-94%" },
  { lo: 0.95, hi: 1.01, etiqueta: "95-99%" }, // límite superior inclusivo de 1.0 (100%)
];

const MUESTRA_MINIMA = 20;
const RESULTADOS_RESUELTOS: Resultado[] = ["GANO", "PERDIO", "EMPATE"];

function bandaDe(confianza: number): string | null {
  const banda = BANDAS.find((b) => confianza >= b.lo && confianza < b.hi);
  return banda ? banda.etiqueta : null;
}

function resumirGrupo(
  grupo: PickCalibracion[],
  banda: string,
  fuenteConfianza: "TODAS" | FuenteConfianza,
): BandaCalibracion {
  const ganadas = grupo.filter((p) => p.resultado === "GANO").length;
  const perdidas = grupo.filter((p) => p.resultado === "PERDIO").length;
  const empates = grupo.filter((p) => p.resultado === "EMPATE").length;
  const decididas = ganadas + perdidas;
  const tasaReal = decididas > 0 ? ganadas / decididas : null;
  const confianzaPromedio = grupo.reduce((acc, p) => acc + p.confianza, 0) / grupo.length;
  const diferencia = tasaReal !== null ? confianzaPromedio - tasaReal : null;

  return {
    banda,
    fuenteConfianza,
    cantidad: grupo.length,
    ganadas,
    perdidas,
    empates,
    confianzaPromedio,
    tasaReal,
    diferencia,
    muestraInsuficiente: grupo.length < MUESTRA_MINIMA,
  };
}

/**
 * Agrupa los picks resueltos en bandas de confianza y compara la
 * confianza promedio declarada contra la tasa real de acierto.
 *
 * Interpretación esperada: si una banda muestra tasaReal muy por debajo de
 * confianzaPromedio (diferencia positiva grande), el sistema está
 * sobreconfiado en esa banda y las confianzas deben ajustarse a la baja.
 *
 * Los empates cuentan en `cantidad` pero se excluyen del denominador de
 * tasaReal (ganadas / (ganadas + perdidas)): un push no confirma ni refuta
 * la confianza asignada.
 *
 * Cada banda se desglosa también por fuenteConfianza (fila 'TODAS',
 * 'CALCULADA' y 'JUICIO'), porque mezclar ambas sin distinguirlas fue el
 * error original que este proyecto corrige.
 *
 * Una banda con menos de 20 picks queda marcada muestraInsuficiente=true;
 * no debe usarse para concluir que esa banda está sobre o subconfiada.
 */
export function reporteCalibracion(picks: PickCalibracion[]): BandaCalibracion[] {
  const resueltos = picks
    .filter((p) => p.resultado !== null && RESULTADOS_RESUELTOS.includes(p.resultado))
    .map((p) => ({ ...p, banda: bandaDe(p.confianza) }))
    .filter((p): p is PickCalibracion & { banda: string } => p.banda !== null);

  const filas: BandaCalibracion[] = [];

  for (const { etiqueta } of BANDAS) {
    const bandaPicks = resueltos.filter((p) => p.banda === etiqueta);
    if (bandaPicks.length === 0) continue;

    filas.push(resumirGrupo(bandaPicks, etiqueta, "TODAS"));

    for (const fuente of ["CALCULADA", "JUICIO"] as const) {
      const subGrupo = bandaPicks.filter((p) => p.fuenteConfianza === fuente);
      if (subGrupo.length === 0) continue;
      filas.push(resumirGrupo(subGrupo, etiqueta, fuente));
    }
  }

  return filas;
}

/**
 * Resultado económico real de los picks resueltos, sin adornos.
 *
 * El desglose `porNivel` reporta cantidad y resultados (GANO/PERDIO/
 * EMPATE) por nivel de pureza (Diamante/Oro/etc.), no montos: una misma
 * boleta física puede combinar patas de distinto nivel, así que repartir
 * el stake/payout de esa boleta entre niveles sería arbitrario.
 *
 * Los montos totales (stake/payout) solo se calculan si al menos un pick
 * resuelto tiene esos datos. Cuando varias patas comparten `ticketId` se
 * asume que el stake/payout de la boleta física está registrado de forma
 * idéntica en cada una de sus filas, así que se deduplica por ticketId
 * antes de sumar para no contar el monto de una misma boleta más de una
 * vez. Filas sin ticketId se tratan como apuestas sueltas.
 */
export function resumenEconomico(picks: PickEconomico[]): ResumenEconomico {
  const resueltos = picks.filter((p) => p.resultado !== null && RESULTADOS_RESUELTOS.includes(p.resultado));

  const porNivel: ResumenEconomico["porNivel"] = {};
  for (const pick of resueltos) {
    if (!porNivel[pick.nivel]) {
      porNivel[pick.nivel] = { cantidad: 0, ganadas: 0, perdidas: 0, empates: 0 };
    }
    const entrada = porNivel[pick.nivel];
    entrada.cantidad++;
    if (pick.resultado === "GANO") entrada.ganadas++;
    else if (pick.resultado === "PERDIO") entrada.perdidas++;
    else if (pick.resultado === "EMPATE") entrada.empates++;
  }

  const conTicket = new Map<string, PickEconomico>();
  const sinTicket: PickEconomico[] = [];
  for (const pick of resueltos) {
    if (pick.ticketId) {
      if (!conTicket.has(pick.ticketId)) conTicket.set(pick.ticketId, pick);
    } else {
      sinTicket.push(pick);
    }
  }
  const dinero = [...conTicket.values(), ...sinTicket];

  // Un hueco explícito es preferible a un número inventado: si ningún pick
  // resuelto tiene stake/payout realmente registrado, reportar null en vez
  // de sumar puros "undefined" y mostrar un engañoso 0 que parecería un
  // resultado económico real.
  const tieneDatosReales = dinero.some((p) => p.stake != null || p.payout != null);

  if (!tieneDatosReales) {
    return {
      totalPicksResueltos: resueltos.length,
      porNivel,
      totalApostado: null,
      totalCobrado: null,
      neto: null,
      advertencia:
        "Ningún pick resuelto tiene stake/payout registrado; no se puede calcular el resultado económico en pesos.",
    };
  }

  const totalApostado = dinero.reduce((acc, p) => acc + (p.stake ?? 0), 0);
  const totalCobrado = dinero.reduce((acc, p) => acc + (p.payout ?? 0), 0);

  return {
    totalPicksResueltos: resueltos.length,
    porNivel,
    totalApostado,
    totalCobrado,
    neto: totalCobrado - totalApostado,
  };
}
